### PR TITLE
Controversial Freedb support

### PR DESCRIPTION
**Pull request by ubitux**
OK so I understand you probably don't want this, but I'm submitting anyway. I'll maintain my crap in my fork in case it's rejected.

The `python-cddb` dependency drop is controversial because I'm actually adding the support to CDDB natively. The main benefit aside from having one less dependency is to have proper unicode support. Note: I should probably move the `main()` to the test suite.

The second commit introduces the `--allow-freedb` option in which we craft fake MusicBrainz metadata. Note that this option is NOT the default, and that it also warns the user about how evil this is.

### CODE REVIEW

**Comment by MerlijnWajer** [comment]
Thanks for working on this. I have no opinion, but wanted to let you know it's appreciated regardless. Let's see what other think, or perhaps have already said.

File: morituri/command/cd.py [hunk 1]
```diff
@@ -279,15 +295,20 @@ def add_arguments(self):
...
         self.parser.add_argument('--disc-template',
             action="store", dest="disc_template",
             default=DEFAULT_DISC_TEMPLATE,
-            help="template for disc file naming (default default)")
+            help="template for disc file naming")
         self.parser.add_argument('-U', '--unknown',
             action="store_true", dest="unknown",
             help="whether to continue ripping if the CD is unknown",
             default=False)
+        self.parser.add_argument('--allow-freedb',
```
Comment (RecursiveForest) [inline morituri/command/cd.py]: How about the short option '-F' for **F**reeDB **F**allback?

File: morituri/command/cd.py [hunk 2]
```diff
@@ -134,10 +134,26 @@ def do(self):
         if not self.program.metadata:
             # fall back to FreeDB for lookup
             cddbid = self.ittoc.getCDDBValues()
-            cddbmd = self.program.getCDDB(cddbid)
-            if cddbmd:
-                sys.stdout.write('FreeDB identifies disc as %s\n' % cddbmd)
 
+            allow_freedb = getattr(self.options, 'allow_freedb', False)
```
Comment (RecursiveForest) [inline morituri/command/cd.py]: there's a default, so I think `allow_freedb = self.options.allow_freedb` is appropriate. Better yet I'd ditch the temporary variable and just use `if self.options.allow_freedb:` on line 147.

File: morituri/command/cd.py [hunk 3]
```diff
@@ -134,10 +134,26 @@ def do(self):
...
+                if allow_freedb:
+                    logger.warning('Using FreeDB metadata, which is not '
+                                   'recommended due to its very low quality '
+                                   'standards.')
+                    logger.warning('You are strongly encouraged to submit CD '
+                                   'information to the MusicBrainz database.')
+                    freedb_data = sfdb.read(match['category'], match['discid'])
+                    self.program.metadata = self.program.craftMusicBrainzFromFreeDB(freedb_data)
+
+        if not self.program.metadata:
```
Comment (RecursiveForest) [inline morituri/command/cd.py]: Following the convention of failing early:

```python
if matches:
    ...
    if not self.options.allow_freedb and not self.options.unknown:
        logger.critical(...)
        return -1
    logger.warning(...)
    freedb_data ...
    ...
```

File: morituri/command/cd.py [hunk 4]
```diff
@@ -134,10 +134,26 @@ def do(self):
...
-            cddbmd = self.program.getCDDB(cddbid)
-            if cddbmd:
-                sys.stdout.write('FreeDB identifies disc as %s\n' % cddbmd)
 
+            allow_freedb = getattr(self.options, 'allow_freedb', False)
+
+            sfdb = simplefreedb.SimpleFreeDB()
+            matches = sfdb.query(cddbid[0], cddbid[1], cddbid[2:-1], cddbid[-1])
+            if matches:
+                match = matches[0] # TODO: honor --prompt
```
Comment (RecursiveForest) [inline morituri/command/cd.py]: If you take a gander at the work I did to rewrite AccurateRip, I factored a generic prompt function you might find useful here.

File: morituri/common/simplefreedb.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+# You should have received a copy of the GNU General Public License
+# along with morituri.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+import socket
+import getpass
+import urllib
+import morituri
+
+class SimpleFreeDB:
```
Comment (RecursiveForest) [inline morituri/common/simplefreedb.py]: I know morituri is organised this way, but since you're not storing any dynamic data on the class, I think this entire module might be a good candidate for a more functional style. It'll read pretty much the same way, but indented 4 characters less and with `self`less signatures.

File: morituri/common/simplefreedb.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+    def _split_dtitle(self, dtitle):
+        # Note: we can not use a simple dtitle.split('/') here because the
+        # slash could be escaped.
+        artist, title = re.split(self._slash_split_regex, dtitle, maxsplit=1)
+        return artist.rstrip(), title.lstrip()
+
+    def _craft_match(self, categ, discid_str, dtitle):
+        artist, title = self._split_dtitle(dtitle)
+        return {'category': categ,
```
Comment (RecursiveForest) [inline morituri/common/simplefreedb.py]: just spell out 'category'

File: morituri/command/cd.py [hunk 7]
```diff
@@ -134,10 +134,26 @@ def do(self):
...
             # fall back to FreeDB for lookup
             cddbid = self.ittoc.getCDDBValues()
-            cddbmd = self.program.getCDDB(cddbid)
-            if cddbmd:
-                sys.stdout.write('FreeDB identifies disc as %s\n' % cddbmd)
 
+            allow_freedb = getattr(self.options, 'allow_freedb', False)
+
+            sfdb = simplefreedb.SimpleFreeDB()
+            matches = sfdb.query(cddbid[0], cddbid[1], cddbid[2:-1], cddbid[-1])
```
Comment (RecursiveForest) [inline morituri/command/cd.py]: Maybe the cddbid parsing logic should be done inside of sfdb.query(), because this is pretty opaque. If you know the sfdb.query() signature, sure you can work it out, but outside of your test cases it doesn't seem like you're ever calling sfdb.query() without having an already composed cddbid.

File: morituri/command/cd.py [hunk 8]
```diff
@@ -279,15 +295,20 @@ def add_arguments(self):
         self.parser.add_argument('--track-template',
             action="store", dest="track_template",
             default=DEFAULT_TRACK_TEMPLATE,
-            help="template for track file naming (default default)")
+            help="template for track file naming")
```
Comment (RecursiveForest) [inline morituri/command/cd.py]: these (and the above whitespace change in getPath) are both good changes, but I think don't belong in this PR.

File: morituri/common/program.py [hunk 9]
```diff
@@ -257,33 +257,24 @@ def getPath(self, outdir, template, mbdiscid, i, profile=None,
...
         ret = os.path.join(outdir, template % v)
-
-
-
         return ret
 
-    def getCDDB(self, cddbdiscid):
-        """
-        @param cddbdiscid: list of id, tracks, offsets, seconds
+    def craftMusicBrainzFromFreeDB(self, freedb_data):
```
Comment (RecursiveForest) [inline morituri/common/program.py]: A short docstring explaining that this is creating musicbrainz-sytle metadata from a freedb entry would make this function perfect.

File: morituri/common/program.py [hunk 10]
```diff
@@ -465,7 +456,9 @@ def getTagList(self, number):
             if self.metadata.release is not None:
                 tags['DATE'] = self.metadata.release
 
-            if number > 0:
+            has_mbinfo = mbidTrack and mbidTrackArtist and mbidAlbum \
+                    and mbidTrackAlbum and mbDiscId
+            if number > 0 and has_mbinfo:
```
Comment (RecursiveForest) [inline morituri/common/program.py]: Would it be worth assigning tags['...TRACKID'] et al. within individual if branches instead of all at once? I'm not sure if it's useful or not to have, say, ARTISTID and not ALBUMID, or if that would ever happen, but it might be worth it? *shrugs*

File: morituri/common/simplefreedb.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+# -*- Mode: Python; test-case-name: morituri.test.test_common_program -*-
+# vi:si:et:sw=4:sts=4:ts=4
+
+# Morituri - for those about to RIP
+
+# Copyright (C) 2017 Clément Bœsch <u@pkh.me>
+
+# This file is part of morituri.
```
Comment (RecursiveForest) [inline morituri/common/simplefreedb.py]: whipper!

File: morituri/common/simplefreedb.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+# -*- Mode: Python; test-case-name: morituri.test.test_common_program -*-
+# vi:si:et:sw=4:sts=4:ts=4
+
+# Morituri - for those about to RIP
```
Comment (RecursiveForest) [inline morituri/common/simplefreedb.py]: RIP morituri-- it's whipper now >:D

File: morituri/common/simplefreedb.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with morituri.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+import socket
+import getpass
+import urllib
```
Comment (RecursiveForest) [inline morituri/common/simplefreedb.py]: I know it requires another dependency (although one that we may introduce for ARv2 support regardless), but it might be worth using the 'requests' library in lieu of urllib here.

File: morituri/common/simplefreedb.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+                    raise Exception('Tracks titles are not monotically incrementing')
+                trackn += 1
+                if not n:
+                    data['tracks'] = []
+                data['tracks'].append(value)
+
+        return data
+
+def main():
+    test_queries = (
```
Comment (RecursiveForest) [inline morituri/common/simplefreedb.py]: I think it would be preferable to decouple testing from the actual code by creating an accompanying test module, like the rest of the application.

**Comment by RecursiveForest** [comment]
Overall this is pretty solid. I left a few notes regarding style and organisation, but the core functionality looks pretty good.

I'm not convinced yet it's worth the potential feature creep, especially because we want to encourage MusicBrainz support, but I can see a case being made for its inclusion. You could also consider writing a small tagging tool on top of simplefreedb that applies freedb-sourced metadata to flacs that were ripped by whipper with the --unknown flag.

Would you consider making simplefreedb into its own independent python module that could be developed independently of whipper, which we would then import? The two existent cddb modules for python are not particularly well packaged (and we know at least one is lacking in UTF-8 support, apparently), so there's definitely a place for it. This is pretty clean code.

File: morituri/command/cd.py (hunk 1 above)
Comment (ubitux) [reply morituri/command/cd.py]: I'd rather not make using this option easy and didn't want to reserve a 1-letter option for a somehow "legacy" mode.

File: morituri/command/cd.py (hunk 8 above)
Comment (ubitux) [reply morituri/command/cd.py]: The whitespace change seems to have been included already by some other change. I'll take out the default-default patch out of it and make a new pull request in a moment.

File: morituri/common/simplefreedb.py (hunk 11 above)
Comment (ubitux) [reply morituri/common/simplefreedb.py]: Yep, I'm in the process of rebasing that stuff, it was morituri when I wrote that patchset. I won't forget to fix that :)

File: morituri/common/simplefreedb.py (hunk 13 above)
Comment (ubitux) [reply morituri/common/simplefreedb.py]: Why? urllib is builtin. I got so many issues with requests...

**Comment by ubitux** [comment]
> Would you consider making simplefreedb into its own independent python module that could be developed independently of whipper, which we would then import?

I don't think I'll be motivated enough to do all the packaging effort, I'm sorry. Feel free to do so though.

Thanks for the long review, I'll make the adjustments you requested soon.

File: morituri/command/cd.py (hunk 2 above)
Comment (ubitux) [reply morituri/command/cd.py]: Good point, I have no idea what pattern I was trying to copy here...

File: morituri/command/cd.py (hunk 4 above)
Comment (ubitux) [reply morituri/command/cd.py]: `git grep -i prompt` doesn't seem to raise anything related. Can you point out where I can find it?

File: morituri/common/simplefreedb.py (hunk 5 above)
Comment (ubitux) [reply morituri/common/simplefreedb.py]: I'm sorry, I think I prefer using the class system for at least namespace. Python is way too much a free4all language already, so it helps with the scoping.

File: morituri/common/simplefreedb.py (hunk 6 above)
Comment (ubitux) [reply morituri/common/simplefreedb.py]: Sure, will do.

File: morituri/command/cd.py (hunk 7 above)
Comment (ubitux) [reply morituri/command/cd.py]: I'd rather have `getCDDBValues()` not build a custom array of stuff. It would indeed be simpler here to just pass `cddbid`, but the way `result` is crafted in `getCDDBValues()` makes me believe it was originally written by someone who had no clue about data structures. If you consider `SimpleFreeDB` could live outside the `whipper` project, it is saner to have this prototype, because the parameters are constructed separately: there is no such thing as "grab all the info and pack them in an array that way" thing outside of `whipper`.

File: morituri/command/cd.py (hunk 8 above)
Comment (ubitux) [reply morituri/command/cd.py]: PR : https://github.com/JoeLametta/whipper/pull/188

File: morituri/common/program.py (hunk 9 above)
Comment (ubitux) [reply morituri/common/program.py]: Sure, will do.

File: morituri/command/cd.py (hunk 1 above)
Comment (RecursiveForest) [reply morituri/command/cd.py]: All options should have a short form.

File: morituri/command/cd.py (hunk 3 above)
Comment (ubitux) [reply morituri/command/cd.py]: I'd like to keep the diff as small as possible, and I'm not sure I'm following what you are trying to restructure here. I'd rather have such change on top of the proposed change if you don't mind.

File: morituri/command/cd.py (hunk 4 above)
Comment (RecursiveForest) [reply morituri/command/cd.py]: Oh, it was @ https://github.com/RecursiveForest/whipper/commit/ca33dcfe79af22e9ee3759659b5473bc517f0746#diff-aa28dbdd5d3fa03298c07270deac1ec1R40 , not part of the AR PR.

File: morituri/common/simplefreedb.py (hunk 5 above)
Comment (RecursiveForest) [reply morituri/common/simplefreedb.py]: I don't understand-- modules provide namespacing.

File: morituri/common/program.py (hunk 10 above)
Comment (ubitux) [reply morituri/common/program.py]: Nope, this `if` is protecting the insert of MusicBrainZ tags, and you absolutely don't want them when using `--allow-freedb`. This code basically checks for every field that is supposed to be present with MusicBrainZ; if any is missing, that's because we're hacking FreeDB stuff in it and so you don't want to end up with broken MusicBrainZ references obtained from dubious sources.

File: morituri/command/cd.py (hunk 1 above)
Comment (ubitux) [reply morituri/command/cd.py]: Many don't (`--cdr`, `--track-template`, `--disc-template` in this module)

File: morituri/common/simplefreedb.py (hunk 5 above)
Comment (ubitux) [reply morituri/common/simplefreedb.py]: Depends how you import it. Anyway, I'm just trying to be consistent with the common way of writing Python. Do you think it's really a problem to wrap an API in a class?

File: morituri/command/cd.py (hunk 4 above)
Comment (ubitux) [reply morituri/command/cd.py]: Oh, but that's not mainline, or is it?

File: morituri/common/simplefreedb.py (hunk 14 above)
Comment (ubitux) [reply morituri/common/simplefreedb.py]: This is not actually a test but more a debug tool (it's not testing output, it's dumping data). I'd prefer to keep tests depending on external websites such as freedb.org outside of the test suite.

File: morituri/command/cd.py (hunk 1 above)
Comment (RecursiveForest) [reply morituri/command/cd.py]: It's something we're planning on fixing shortly, although I understand that you may not be privy to transient information on IRC.

Comment (RecursiveForest) [reply morituri/command/cd.py]: I also can't argue there's a consensus on this topic, I just strongly prefer BSD-style interfaces; so do what you want.

**Comment by anarcat** [comment]
i would argue for this even if only for archival purposes. here i'm using whipper to rescue an old personal CD collection and i happened to have tagged some of those disks on freeDB. they have very low distribution (this is probably the only copy left in the universe) so there's little point in repopulating musicbrainz with this stuff...

**Comment by JoeLametta** [comment]
@anarcat Well, that's a pretty niche use case, isn't it? (considering that you can also tag those files after whipper has finished its job)

Anyway, right now we aren't using the `python-cddb` module anymore (see https://github.com/JoeLametta/whipper/pull/276 for more details), so what's left relevant in this pull request is the second commit ([cd: add --allow-freedb option](https://github.com/JoeLametta/whipper/pull/138/commits/4d582d3776bf5b1c8ca8b61b2da7ee2cd80d72c0)). That one may a worthful addition: let's keep discussing the pros and cons of this proposal...

**Comment by anarcat** [comment]
just to weigh another point in - what i ended up doing here is to feed the data back into musicbrainz. there are two ways of viewing this:

 1. yay, people are contributing
 2. crap, people are feeding useless old, bad data from freedb into musicbrainz still?

i'd rather not encourage the second pattern, but it's true that we might not encourage people to contribute to musicbrainz if data is only on freedb....

**Comment by JoeLametta** [comment]
Related to #44.

**Comment by peioe** [comment]
It would be really nice to add an option to allow grabbing metadata from freedb. It's completely ok if it doesn't by default but it would really help with the adoption of whipper imo.

**Comment by JoeLametta** [comment]
@Freso @MerlijnWajer What's your opinion about [this](https://github.com/whipper-team/whipper/pull/138#issuecomment-427561631)?

**Comment by Freso** [comment]
Considering that [freedb.org](http://www.freedb.org/) is sunsetting and seems like it will be shut down in a few months, I vote to reject this PR and any other attempt of freedb compatibility. Even if it gets resurrected (on the same domain or elsewhere), the technology behind it is ancient and largely in maintenance-only mode (if even that much!). Let’s move on.

billede

**Comment by JoeLametta** [comment]
@Freso Thanks for the reply: didn't know about the imminent shutdown! For this reason I'm closing the pull request.

**Comment by peioe** [comment]
That's sad, there are a lot more releases on freedb than MB, and if someone hosts a mirror (and they already exist) it could still be used. In my experience having only MB is pretty much equivalent to having no metadata at all unfortunately.

**Comment by ubitux** [comment]
> That's sad, there are a lot more releases on freedb than MB, and if someone hosts a mirror (and they already exist) it could still be used. In my experience having only MB is pretty much equivalent to having no metadata at all unfortunately.

That was pretty much my experience and the reason I hacked this branch. Out of all the CDs I tried to RIP with whipper (about 20), almost none where found on MB, while almost all of them where on freedb.

**Comment by anarcat** [comment]
Same here: what brought me here was the daunting task of ripping my old CD collection out of fear it would dissolve itself in the mists of time (which it partly did). Because many of those releases predate the very existence of MB, a lot of those releases are only on FreeDB.org... 

But I guess if the site dies, it's not worth keeping this implementation around.

@peioe which mirror are you aware of?

**Comment by MerlijnWajer** [comment]
Here's another controversial thought - what about adding gracenote support? They pretty much "inherited" their data from FreeDB.org anyway. (I might already have code for this)

**Comment by Freso** [comment]
I’m principally opposed to supporting a closed source, proprietary data source. (Also, Gracenote were the ones responsible for shutting off contributors’ access to CDDB 20 years ago, leading to the creation of FreeDB, MusicBrainz, et al.)

**Comment by peioe** [comment]
> @peioe which mirror are you aware of?

http://www.freedb.org/en/faq.3.html#18
http://www.freedb.org/en/faq.3.html#112
http://www.freedb.org/en/download__server_software.4.html
http://www.freedb.org/en/download__database.10.html

EAC for example allows you to specify the url to a freedb server, so you can just host your own.

Even if freedb goes down tomorrow, the data it collected over the years is still extremely useful, and if it's possible to host a mirror it would be great to be able to use it with whipper.
